File: java/02_integrate_agent_with_llm/agent-tests/src/test/java/com/workshop/agent/tests/AgentTests.java
package com.workshop.agent.tests;

import com.workshop.agent.application.*;
import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
public class AgentTests {
    private final UserInput input = new InputStub("Hello, Agent!");
    private final DisplayStub display = new DisplayStub();

    @Test
    public void shows_user_input_on_display() {
        LanguageModel model = new RepeatingLanguageModel();
        Agent agent = new Agent(input, model, display);

        agent.run();

        assertThat(display.content).startsWith("User: Hello, Agent!\n");
    }

    @Test
    public void sends_user_input_to_model() {
        RepeatingLanguageModel model = new RepeatingLanguageModel();
        Agent agent = new Agent(input, model, display);

        agent.run();

        assertThat(model.capturedPrompts).hasSize(1);
        assertThat(model.capturedPrompts.get(0)).containsExactly(
                new Message("user", "Hello, Agent!")
        );
    }

    @Test
    public void shows_model_response_on_display() {
        LanguageModel model = new RepeatingLanguageModel();
        Agent agent = new Agent(input, model, display);

        agent.run();

        assertThat(display.content).isEqualTo(
                """
                        User: Hello, Agent!
                        Assistant: You said: "Hello, Agent!"
                        """
        );
    }
}

class RepeatingLanguageModel implements LanguageModel {
    public final List<List<Message>> capturedPrompts = new ArrayList<>();

    @Override
    public Message prompt(List<Message> messages) {
        capturedPrompts.add(new ArrayList<>(messages));
        Message lastMessage = messages.get(messages.size() - 1);
        return new Message("assistant", "You said: \"" + lastMessage.content() + "\"");
    }
}

class DisplayStub implements Display {
    public String content = "";

    @Override
    public void show(Message message) {
        String role = message.role();
        String capitalizedRole = Character.toUpperCase(role.charAt(0)) + role.substring(1);
        content += capitalizedRole + ": " + message.content() + "\n";
    }
}

class InputStub implements UserInput {
    private final String message;

    public InputStub(String message) {
        this.message = message;
    }

    @Override
    public String getInput() {
        return message;
    }
}

File: csharp/04_explain_how_to_use_the_tool/Agent/Agent.Tests/AgentTests.cs
using Agent.Application;

namespace Agent.Tests;

public class AgentTests
{
    [Test]
    public void UserInput_Shown_On_Display()
    {
        var input = new InputStub(["Hello, Agent!", ""]);
        var model = new LanguageModelSpy();
        var display = new DisplayStub();
        var agent = new Application.Agent(input, model, display);

        agent.Run();

        Assert.That(display.Content, Does.StartWith("User: Hello, Agent!\n"));
    }

    [Test]
    public void UserInput_Sent_To_Model()
    {
        var input = new InputStub(["Hello, Agent!", ""]);
        var model = new LanguageModelSpy();
        var display = new DisplayStub();
        var agent = new Application.Agent(input, model, display);

        agent.Run();

        Assert.That(model.CapturedPrompts, Does.Contain(
                new List<Message> {new("user", "Hello, Agent!")}
        ));
    }

    [Test]
    public void LanguageModelResponse_Shown_On_Display()
    {
        var input = new InputStub(["Hello, Agent!", ""]);
        var model = new LanguageModelStub("Hello, what can I do for you, today!");
        var display = new DisplayStub();
        var agent = new Application.Agent(input, model, display);

        agent.Run();

        Assert.That(display.Content, Does.StartWith(
            "User: Hello, Agent!\n" +
            "Assistant: Hello, what can I do for you, today!\n"
        ));
    }
    
    [Test]
    public void Displays_Back_And_Forth_Chat()
    {
        var input = new InputStub(["Hello, Agent!", "I have another Message for you.", ""]);
        var model = new RepeatingLanguageModel();
        var display = new DisplayStub();
        var agent = new Application.Agent(input, model, display);

        agent.Run();

        Assert.That(display.Content, Is.EqualTo(
            "User: Hello, Agent!\n" +
            "Assistant: You said: \"Hello, Agent!\"\n" +
            "User: I have another Message for you.\n" +
            "Assistant: You said: \"I have another Message for you.\"\n"
        ));
    }
    
    [Test]
    public void Sends_Whole_Context_To_Model()
    {
        var input = new InputStub(["Hello, Agent!", "I have another Message for you.", ""]);
        var model = new LanguageModelSpy();
        var display = new DisplayStub();
        var agent = new Application.Agent(input, model, display);

        agent.Run();

        Assert.That(model.CapturedPrompts, Is.EqualTo(new List<List<Message>>
        {
            new() { new Message("user", "Hello, Agent!") },
            new()
            {
                new Message("user", "Hello, Agent!"),
                new Message("assistant", "You said: \"Hello, Agent!\""),
                new Message("user", "I have another Message for you."),
            }
        }));
    }
}

public class RepeatingLanguageModel : ILanguageModel
{
    public Message Prompt(IEnumerable<Message> messages)
    {
        return new Message("assistant", "You said: \"" + messages.Last().Content + "\"");
    }
}

public class LanguageModelStub(string message) : ILanguageModel
{
    public Message Prompt(IEnumerable<Message> messages)
    {
        return new Message("assistant", message);
    }
}

public class LanguageModelSpy : ILanguageModel
{
    public readonly List<List<Message>> CapturedPrompts = [];

    public Message Prompt(IEnumerable<Message> messages)
    {
        CapturedPrompts.Add(messages.ToList());
        return new Message("assistant", "You said: \"" + messages.Last().Content + "\"");
    }
}

public class DisplayStub : IDisplay
{
    public string Content = "";

    public void Show(string text)
    {
        Content += text + "\n";
    }
}

public class InputStub(List<string> message) : IUserInput
{
    public string GetInput()
    {
        var first = message.First();
        if (message.Count > 1)
        {
            message.RemoveAt(0);
        }

        return first;
    }
}